bes_publisher/buildevent/service.go

```diff
@@ -248,18 +263,71 @@ func (b *buildStream) updateAttrs(event *bes.BuildEvent) {
 func (b *buildStream) maybePublish(event *bes.BuildEvent) error {
 	copy := &bes.BuildEvent{Id: event.Id, Payload: event.Payload}
 
-	switch event.Payload.(type) {
+	extraAttrs := map[string]string{}
+	switch payload := event.Payload.(type) {
 	default:
 		metricBuildEventServiceEventCount.WithLabelValues(oneofType(event.Payload), "dropped").Inc()
 		return nil
+	case *bes.BuildEvent_Configured:
+		// To date, nothing really cares about this message directly. However, we do
+		// care about TargetComplete messages; specifically, being able to tell what
+		// kind of rule it was. TargetComplete has deprecated fields for this, and
+		// says to look at TargetConfigured events for the canonical info. So, we
+		// need to remember target types here for each target, to pair them with
+		// TargetComplete events later.
+		eventID := event.GetId().GetTargetConfigured()
+		k := targetKey(eventID.GetLabel(), eventID.GetAspect())
+		targetType := payload.Configured.GetTargetKind()
+		// These strings look like: `py_binary rule`
+		// Strip off ` rule` so that downstream code is more sensible
+		if strings.HasSuffix(targetType, " rule") {
+			b.typeFromLabelAndAspect[k] = strings.TrimSuffix(targetType, " rule")
+		} else {
+			// If it doesn't have this suffix, we actually have no idea what the
+			// string looks like. Add logging here if this is a problem.
+			metricUnknownTargetType.WithLabelValues("unknown_rule_str_format").Inc()
+			b.typeFromLabelAndAspect[k] = targetType
+		}
+		// Mark this event as otherwise unhandled
+		metricBuildEventServiceEventCount.WithLabelValues(oneofType(event.Payload), "recorded_only").Inc()
+		return nil
 	case *bes.BuildEvent_Started:
 	case *bes.BuildEvent_BuildMetadata:
 	case *bes.BuildEvent_WorkspaceStatus:
+	case *bes.BuildEvent_Completed:
+		// Look up the type of this target (see above in handling TargetConfigured)
+		// and stuff this as an attr for just this message.
+		eventID := event.GetId().GetTargetCompleted()
+		k := targetKey(eventID.GetLabel(), eventID.GetAspect())
+		targetType, ok := b.typeFromLabelAndAspect[k]
+		if ok {
+			extraAttrs["rule_type"] = targetType
+		} else {
+			metricUnknownTargetType.WithLabelValues("unmatched_completed_target").Inc()
+		}
+
+		// Annoyingly, subsequent events can't be identified by label + aspect, so
+		// remember a label + configuration as well to match up target kinds.
+		k = targetKey(eventID.GetLabel(), eventID.GetConfiguration().GetId())
+		b.typeFromLabelAndConfiguration[k] = targetType
```

Comment:
Is it intended to put the default value in here if it isnt found in the map on line 302?

Reply:
No - moved this into the "ok" block